spec/12-the-scala-standard-library.md

```diff
@@ -642,6 +642,12 @@ object Predef {
   def classOf[T]: Class[T] = null
    // this is a dummy, classOf is handled by compiler.
 
+  // valueOf -----------------------------------------------------------
+
+  /** Retrieve the single value of a type with a unique inhabitant. */
+  @inline def valueOf[T](implicit vt: ValueOf[T]): T {} = vt.value
```

Comment:
@soc I disagree with your reasoning. The fact that `classOf[T]` is a builtin instead of taking an implicit `ClassTag[T]` is an artifact of history. Were we to define `classOf` and `ClassTag` *today*, we would make the implicit materialization of `ClassTag` the only builtin, and implement `classOf` in user-space as
```scala
def classOf[T](implicit ct: ClassTag[T]): Class[T] = ct.runtimeClass
```
This would also have a nice benefit that we could simply write `classOf[T]` instead of `classTag[T].runtimeClass` when we have an implicit `ClassTag[T]` in scope.

Similarly, were `new Array[T](len)` be defined *today*, it would have an implicit `ClassTag[T]` as well.

Since we are defining `ValueOf` today, we can do it the right way, with the implicit materialization be the only builtin, and `valueOf[T]` being purely user-space.

I do not understand your argument of "bullet-proof". If I have a `v: ValueOf[T]` at my disposal, there's nothing wrong in passing it explicitly to `valueOf[T](v)`. The type of `v` already guarantees that `v` is the right `ValueOf[T]` to be passed to `valueOf[T]` anyway.